core/run.go

```diff
@@ -0,0 +1,222 @@
+package core
+
+import (
+	"context"
+	"errors"
+	"net/http"
+	"time"
+
+	"chain/core/accesstoken"
+	"chain/core/account"
+	"chain/core/asset"
+	"chain/core/config"
+	"chain/core/fetch"
+	"chain/core/generator"
+	"chain/core/leader"
+	"chain/core/pin"
+	"chain/core/query"
+	"chain/core/rpc"
+	"chain/core/txbuilder"
+	"chain/core/txdb"
+	"chain/core/txfeed"
+	"chain/database/pg"
+	"chain/log"
+	"chain/protocol"
+	"chain/protocol/bc"
+)
+
+const (
+	blockPeriod              = time.Second
+	expireReservationsPeriod = time.Second
+)
+
+// RunOption describes a runtime configuration option.
+type RunOption func(*API)
+
+// AlternateAuth configures the Core to use authFn to authenticate
+// incoming requests in addition to the default access token authentication.
+func AlternateAuth(authFn func(*http.Request) bool) RunOption {
+	return func(a *API) { a.altAuth = authFn }
+}
+
+// BlockSigner configures the Core to use signFn to handle block-signing
+// requests. In production, this will be a function to call out to signerd
+// and its HSM. In development, it'll use the MockHSM.
+func BlockSigner(signFn func(context.Context, *bc.Block) ([]byte, error)) RunOption {
+	return func(a *API) { a.signer = signFn }
+}
+
+// IndexTransactions configures whether or not transactions should be
+// annotated and indexed for the query engine.
+func IndexTransactions(b bool) RunOption {
+	return func(a *API) { a.indexTxs = b }
+}
+
+// RateLimit adds a rate-limiting restriction, using keyFn to extract the
+// key to rate limit on. It will allow up to burst requests in the bucket
+// and will refill the bucket at perSecond tokens per second.
+func RateLimit(keyFn func(*http.Request) string, burst, perSecond int) RunOption {
+	return func(a *API) {
+		a.requestLimits = append(a.requestLimits, requestLimit{
+			key:       keyFn,
+			burst:     burst,
+			perSecond: perSecond,
+		})
+	}
+}
+
+// LocalGenerator configures the launched Core to run as a Generator.
+func LocalGenerator(gen *generator.Generator) RunOption {
+	return func(a *API) {
+		if a.remoteGenerator != nil {
+			panic("core configured with local and remote generator")
+		}
+		a.generator = gen
+		a.submitter = gen
+	}
+}
+
+// RemoteGenerator configures the launched Core to fetch blocks from
+// the provided remote generator.
+func RemoteGenerator(client *rpc.Client) RunOption {
```

Comment:
What do you think about GeneratorLocal and GeneratorRemote so they sort together in the godoc output? Nbd tho.